### PR TITLE
[feature](multi-catalog) support hive metastore more events

### PR DESCRIPTION
**Pull request by zddr**
# Proposed changes

Issue Number: close #xxx

## Problem summary

support hive metastore more events

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [x] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [x] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 36.05 seconds
 load time: 645 seconds
 storage size: 17123668228 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230107130805_clickbench_pr_75392.html

File: fe/fe-core/src/main/java/org/apache/doris/datasource/CatalogMgr.java [hunk 1]
```diff
@@ -583,11 +605,251 @@ public void replayDropExternalTable(ExternalObjectLog log) {
...
+        log.setTableName(tableName);
+        log.setTableId(Env.getCurrentEnv().getNextId());
+        replayCreateExternalTable(log);
+        Env.getCurrentEnv().getEditLog().logCreateExternalTable(log);
+    }
+
+    public void replayCreateExternalTable(ExternalObjectLog log) {
+        LOG.debug("ReplayCreateExternalTable,catalogId:[{}],dbId:[{}],tableId:[{}],tableName:[{}]", log.getCatalogId(),
+                log.getDbId(), log.getTableId(), log.getTableName());
+        ExternalCatalog catalog = (ExternalCatalog) idToCatalog.get(log.getCatalogId());
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/CatalogMgr.java:659]: Check if `catalog` is null, just return.
Same for the following `db`

File: fe/fe-core/src/main/java/org/apache/doris/datasource/CatalogMgr.java [hunk 2]
```diff
@@ -583,11 +605,251 @@ public void replayDropExternalTable(ExternalObjectLog log) {
...
+        replayDropExternalDatabase(log);
+        Env.getCurrentEnv().getEditLog().logDropExternalDatabase(log);
+    }
+
+    public void replayDropExternalDatabase(ExternalObjectLog log) {
+        writeLock();
+        try {
+            LOG.debug("ReplayDropExternalTable,catalogId:[{}],dbId:[{}],tableId:[{}]", log.getCatalogId(),
+                    log.getDbId(), log.getTableId());
+            ExternalCatalog catalog = (ExternalCatalog) idToCatalog.get(log.getCatalogId());
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/CatalogMgr.java:695]: Ditto

File: fe/fe-core/src/main/java/org/apache/doris/datasource/CatalogMgr.java [hunk 3]
```diff
@@ -583,11 +605,251 @@ public void replayDropExternalTable(ExternalObjectLog log) {
...
+        replayCreateExternalDatabase(log);
+        Env.getCurrentEnv().getEditLog().logCreateExternalDatabase(log);
+    }
+
+    public void replayCreateExternalDatabase(ExternalObjectLog log) {
+        writeLock();
+        try {
+            LOG.debug("ReplayCreateExternalDatabase,catalogId:[{}],dbId:[{}],dbName:[{}]", log.getCatalogId(),
+                    log.getDbId(), log.getDbName());
+            ExternalCatalog catalog = (ExternalCatalog) idToCatalog.get(log.getCatalogId());
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/CatalogMgr.java:730]: Ditto

File: fe/fe-core/src/main/java/org/apache/doris/datasource/CatalogMgr.java [hunk 4]
```diff
@@ -583,11 +605,251 @@ public void replayDropExternalTable(ExternalObjectLog log) {
...
+        log.setTableId(table.getId());
+        log.setPartitionNames(partitionNames);
+        replayAddExternalPartitions(log);
+        Env.getCurrentEnv().getEditLog().logAddExternalPartitions(log);
+    }
+
+    public void replayAddExternalPartitions(ExternalObjectLog log) {
+        LOG.debug("ReplayAddExternalPartitions,catalogId:[{}],dbId:[{}],tableId:[{}]", log.getCatalogId(),
+                log.getDbId(), log.getTableId());
+        ExternalCatalog catalog = (ExternalCatalog) idToCatalog.get(log.getCatalogId());
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/CatalogMgr.java:768]: Ditto

File: fe/fe-core/src/main/java/org/apache/doris/datasource/CatalogMgr.java [hunk 5]
```diff
@@ -583,11 +605,251 @@ public void replayDropExternalTable(ExternalObjectLog log) {
...
+        log.setTableId(table.getId());
+        log.setPartitionNames(partitionNames);
+        replayDropExternalPartitions(log);
+        Env.getCurrentEnv().getEditLog().logDropExternalPartitions(log);
+    }
+
+    public void replayDropExternalPartitions(ExternalObjectLog log) {
+        LOG.debug("ReplayDropExternalPartitions,catalogId:[{}],dbId:[{}],tableId:[{}]", log.getCatalogId(),
+                log.getDbId(), log.getTableId());
+        ExternalCatalog catalog = (ExternalCatalog) idToCatalog.get(log.getCatalogId());
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/CatalogMgr.java:806]: Ditto

File: fe/fe-core/src/main/java/org/apache/doris/datasource/CatalogMgr.java [hunk 6]
```diff
@@ -583,11 +605,251 @@ public void replayDropExternalTable(ExternalObjectLog log) {
...
+        log.setTableId(table.getId());
+        log.setPartitionNames(partitionNames);
+        replayRefreshExternalPartitions(log);
+        Env.getCurrentEnv().getEditLog().logInvalidateExternalPartitions(log);
+    }
+
+    public void replayRefreshExternalPartitions(ExternalObjectLog log) {
+        LOG.debug("replayRefreshExternalPartitions,catalogId:[{}],dbId:[{}],tableId:[{}]", log.getCatalogId(),
+                log.getDbId(), log.getTableId());
+        ExternalCatalog catalog = (ExternalCatalog) idToCatalog.get(log.getCatalogId());
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/CatalogMgr.java:845]: Ditto

File: fe/fe-core/src/main/java/org/apache/doris/datasource/hive/event/AlterDatabaseEvent.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+    protected static List<MetastoreEvent> getEvents(NotificationEvent event,
+            String catalogName) {
+        return Lists.newArrayList(new AlterDatabaseEvent(event, catalogName));
+    }
+
+    @Override
+    protected void process() throws MetastoreNotificationException {
+        // only can change properties,we do nothing
+        infoLog("catalogName:[{}],dbName:[{}]", catalogName, dbName);
```
[the remark this answers is not in the document]
Comment (zddr) [reply fe/fe-core/src/main/java/org/apache/doris/datasource/hive/event/AlterDatabaseEvent.java:46]: current api version we used can not support we do this

File: fe/fe-core/src/main/java/org/apache/doris/datasource/CatalogMgr.java [hunk 8]
```diff
@@ -544,6 +544,28 @@ public void replayRefreshExternalDb(ExternalObjectLog log) {
         }
     }
 
+    public void refreshExternalTable(String dbName, String tableName, String catalogName) throws DdlException {
```
[the remark this answers is not in the document]
Comment (zddr) [reply fe/fe-core/src/main/java/org/apache/doris/datasource/CatalogMgr.java:547]: That method is not abstract enough. I abstracted a method and put it in CatalogMgr, and then deleted the method in RefreshMgr

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**